### DISCUSSION TITLE
How to implement stored subgraphs?

### DISCUSSION THREAD

**Discussion by Vithanco** [graphology/graphology#435] [Ideas]
I am wondering whether there is a way to maintain subgraphs as part of the Graph. The subgraphs will be used for graphical representation (see for example [Graphviz Clusters](https://graphviz.org/Gallery/directed/cluster.html). I would like to store subgraphs as part of the graph, so they could be serialised (import and export). Ideally, they should be updated if nodes are dropped. 

Did anybody try do do this? Any suggestions on how to implement this? Many thanks in advance!

**Comment by Yomguithereal** [comment]
@Vithanco I thinks there are two different ways to do this:

1. Creating something like a SubgraphView of a given graph, that masquerades as a proper `graphology` implementation but that will reiplement methods such as `#.forEachNode` etc. by taking into account the fact that only some nodes belong to the subgraph etc. This solution is good if you want to avoid spending too much memory but is not good if you want to optimize time complexity of the subgraph methods. This also means you might have to reimplement a lot of methods to make your view work for you since graphology has quite a wide API. One day I will implement a correct complete subgraph view but I lack the incentives to do so since I never had the use for it.
2. Another solution is to create a synced subgraph index that is basically another graph instance containing the subgraph's data (that you can first create using `graphology-utils/operators/subgraph`. You can then sync it to the parent graph by listening to its [events](https://graphology.github.io/events.html) such as `nodeDropped` etc.

In any case, if you really want to tie your subgraph to a graphology instance for serialization purposes and such, you can add them to a [graph attribute](https://graphology.github.io/attributes.html#graph-attributes) containing a list of your stored subgraphs for instance.
